ARROW-14702: [Doc][C++] Document threading model

westonpace opened the pull request.

westonpace commented:
Todo:

- [ ] Figure out `arrow::internal::Executor`
- [ ] Review documentation for Future methods
- [ ] Figure out why Sphinx is complaining about duplicate definitions for Future
- [ ] Link in ARROW_IO_THREADS env var

westonpace commented:
@pitrou I can't include API docs for `arrow::internal::Executor` because doxygen excludes everything in the internal namespace.  Would it be better to remove all references to the executor (and simply reference the management methods like setting and retrieving capacity) or should we move `Executor` out of the `internal` namespace?

pitrou commented:
I don't know, do we actually want to document `Executor` publicly or just the more general concepts around our execution model?

westonpace commented:
I reduced the scope to remove references to `Executor`.  I left a small blurb in about futures (mainly so I could say "any method that returns a future probably has a synchronous variant")

pitrou commented on docs/source/cpp/getting_started.rst:
```diff
@@ -27,6 +27,7 @@ User Guide
    conventions
    build_system
    memory
+   threading
```
I'd move this much farther below in the table of contents, since this is more of an advanced topic that only comes up when using certain APIs. What do you think?

westonpace replied:
I moved it down to just past the gdb section.

pitrou commented on docs/source/cpp/threading.rst:
```diff
@@ -0,0 +1,100 @@
+   :doc:`Thread management API reference <api/thread>`
+
+Thread Pools
+=======================
```
Be careful, the number of sigils here must be the same as the title length above, otherwise it's incorrect reST syntax (there's probably a warning somewhere in the doc build logs).

westonpace replied:
Fixed.  Thanks!

pitrou commented on docs/source/cpp/threading.rst:
```diff
@@ -0,0 +1,100 @@
+Thread Pools
+=======================
+
+Many Arrow operations distribute work across multiple threads to take
```
I would strive to always say "Arrow C++" when talking about the implementation not the format.

westonpace replied:
Done.

pitrou commented on docs/source/cpp/threading.rst:
```diff
@@ -0,0 +1,100 @@
+
+Many Arrow operations distribute work across multiple threads to take
+advantage of underlying hardware parallelism.  For example, when reading a
+parquet file we can decode each column in parallel.  To achieve this we
```
Nit: "Parquet" capitalized, also let's cross-reference to the corresponding doc page?

westonpace replied:
I'm not sure exactly where the best place would be to cross-reference.  I added a reference to https://arrow.apache.org/docs/cpp/api/formats.html#_CPPv4N7parquet5arrow10FileReader15set_use_threadsEb

pitrou commented on docs/source/cpp/threading.rst:
```diff
@@ -0,0 +1,100 @@
+filesystem such as S3, it is often possible to benefit from many concurrent reads
+and it may be possible to increase I/O performance by increasing the size of the
+I/O thread pool.  The size of the default I/O thread pool can be managed with
+the :ref:`ARROW_IO_THREADS<env_arrow_io_threads>` environment variable or
```
Can simply use the ```:envvar:`ARROW_IO_THREADS` ``` markup (just like for classes, functions, etc.).

westonpace replied:
Thanks, I changed to this and removed the labels.

pitrou commented:
I really like this change, thanks for doing this!

westonpace commented:
@pitrou Thanks for the review.  I think I've addressed the feedback.  I also realized this information applies to python and R as well so I plan on adding a small blurb to those implementations as well.  I also added a small reference in the filesystems page.

pitrou commented:
I'm going to merge this, feel free to open new PR for Python and R additions.
